### PR TITLE
Don't push offset to leaf layers

### PR DESCRIPTION
**Pull request by liyuqian**
For retained rendering, we don't want to push the offset down to each leaf layer. Otherwise, changing an offset layer on the very high level could cascade the change to too many leaves, which means that we can't retain them.

To not push the offset downwards, we simply push a TransformLayer when there's an offset. Skia has a fast path for concatenating scale/translation-only matrix so this operation should be fast (no performance regression is measured on Moto G4).

This also makes our compositor much cleaner as we don't have to pass `layerOffset` all over the places in `XXXLayer::addToScene`.

We can also remove the `offset_` field and its related operations from `flow::PictureLayer, flow::TextureLayer, flow::ChildSceneLayer` in flutter/engine once this change landed.

This is our first step towards https://github.com/flutter/flutter/issues/21756

### CODE REVIEW

**Comment by liyuqian** [comment]
CC @yjbanov : the offset will no longer be accumulated to the leaf layer. Hopefully that makes the web flutter a little easier to implement.

**Comment by yjbanov** [comment]
Have you considered using `Layer.owner` as a marker for retaining layers? I have a working prototype of this if you want to see it.

**Comment by liyuqian** [comment]
@yjbanov Sure, I'll have a look. Can you point me to your prototype?

File: packages/flutter/lib/src/rendering/layer.dart [hunk 1]
```diff
@@ -104,10 +104,7 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   S find<S>(Offset regionOffset);
 
   /// Override this method to upload this layer to the engine.
-  ///
-  /// The `layerOffset` is the accumulated offset of this layer's parent from the
-  /// origin of the builder's coordinate system.
-  void addToScene(ui.SceneBuilder builder, Offset layerOffset);
+  void addToScene(ui.SceneBuilder builder);
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/layer.dart]: this is a breaking change.

**Comment by Hixie** [comment]
I think we should continue to support the old API, even if we change the default implementation of OffsetLayer to use transforms and pass through zero instead of passing through the offset.

File: packages/flutter/lib/src/rendering/layer.dart [hunk 2]
```diff
@@ -560,8 +556,23 @@ class OffsetLayer extends ContainerLayer {
...
+
+  @override
+  void addToScene(ui.SceneBuilder builder) {
+    // Skia has a fast path for concatenating scale/translation only matrices.
+    // Hence this operation should be fast. For retained rendering, we don't
+    // want to push the offset down to each leaf node. Otherwise, changing an
+    // offset layer on the very high level could cascade the change to too many
+    // leaves.
+    final Matrix4 matrix = new Matrix4.translationValues(offset.dx, offset.dy, 0.0);
+    builder.pushTransform(matrix.storage);
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/layer.dart]: I think I'd prefer a dedicated `pushOffset` method on `SceneBuilder` that matches the API of `OffsetLayer`:

```dart
@override
void addToScene(ui.SceneBuilder builder) {
  builder.pushOffset(offset.dx, offset.dy);
}
```

The `dx` and `dy` can be easily (and more efficiently) converted to a transform matrix by the engine. Knowing the difference between a transform and an offset I can use it to collapse a series of offsets into a single offset as an optimization that reduces the total number of layers.

Comment (liyuqian) [reply packages/flutter/lib/src/rendering/layer.dart]: Sounds good to me. https://github.com/flutter/engine/pull/6349 is uploaded to handle this. Once that landed, I can change this (either in this PR, or in the next PR).

**Comment by liyuqian** [comment]
@Hixie Added `[Offset layerOffset = Offset.zero]` back so it won't break the API but will encourage to not use layerOffset.

File: packages/flutter/lib/src/rendering/layer.dart [hunk 3]
```diff
@@ -1363,8 +1378,11 @@ class FollowerLayer extends ContainerLayer {
   void applyTransform(Layer child, Matrix4 transform) {
     assert(child != null);
     assert(transform != null);
-    if (_lastTransform != null)
+    if (_lastTransform != null) {
       transform.multiply(_lastTransform);
+    } else {
+      transform.multiply(Matrix4.translationValues(unlinkedOffset.dx, unlinkedOffset.dy, .0));
+    }
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/layer.dart]: looks like this is fixing an old bug, can you add a test for this?

**Comment by Hixie** [comment]
LGTM.

I'm surprised by the premise of this PR, though. why would we need to send anything to the engine but the new offsets on the leaves?

**Comment by yjbanov** [comment]
@Hixie here's an example:

Frame 1:

```
offset(0, 0)
  opacity()
    offset(10, 10)
      picture()
    pop()
  pop()
pop()
```

Frame 2:

```
offset(10, 10)
  opacity()
    offset(10, 10)
      picture()
    pop()
  pop()
pop()
```

Today, this is submitted to the Engine as:

Frame 1:

```
opacity()
  picture() with offset(10, 10)
pop()
```

Frame 2:

```
opacity()
  picture() with offset(20, 20)
pop()
```

Note that we mutated the tree under the "opacity" node.

After this change, we will submit the layers to the engine unchanged, i.e. the sub-tree rooted at "opacity" is not mutated from frame 1 to frame 2, which tells the engine "nothing's changed under opacity and it is safe to cache it".

**Pull request merged**